teku/src/main/java/tech/pegasys/teku/cli/util/DatabaseMigrater.java

```
@@ -0,0 +1,256 @@
+import org.apache.commons.io.FileUtils;
+import org.hyperledger.besu.metrics.noop.NoOpMetricsSystem;
+import tech.pegasys.teku.cli.options.DataStorageOptions;
+import tech.pegasys.teku.cli.options.Eth2NetworkOptions;
+import tech.pegasys.teku.cli.options.ValidatorClientDataOptions;
+import tech.pegasys.teku.service.serviceutils.layout.DataDirLayout;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.storage.server.DatabaseVersion;
+import tech.pegasys.teku.storage.server.VersionedDatabaseFactory;
+import tech.pegasys.teku.storage.server.kvstore.KvStoreDatabase;
+
+public class DatabaseMigrater {
+  private final DataDirLayout dataDirLayout;
+  private final Consumer<String> statusUpdater;
+  private final DataStorageOptions dataStorageOptions;
+  private final Eth2NetworkOptions eth2NetworkOptions;
+  private final int batchSize;
+  private Optional<Path> migratedDatabasePath = Optional.empty();
+  private final Spec spec;
+  private KvStoreDatabase originalDatabase;
+
+  KvStoreDatabase getOriginalDatabase() {
+    return originalDatabase;
+  }
+
+  KvStoreDatabase getNewDatabase() {
+    return newDatabase;
+  }
+
+  private KvStoreDatabase newDatabase;
+
+  private DatabaseMigrater(
+      final DataDirLayout dataDirLayout,
+      final DataStorageOptions dataStorageOptions,
+      final Eth2NetworkOptions eth2NetworkOptions,
+      final Spec spec,
+      final int batchSize,
+      final Consumer<String> statusUpdater) {
+    this.dataDirLayout = dataDirLayout;
+    this.dataStorageOptions = dataStorageOptions;
+    this.eth2NetworkOptions = eth2NetworkOptions;
+    this.spec = spec;
+    this.batchSize = batchSize;
+    this.statusUpdater = statusUpdater;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public void migrateDatabase(
+      final DatabaseVersion sourceDatabaseVersion, final DatabaseVersion targetDatabaseVersion)
+      throws DatabaseMigraterError {
+    if (migratedDatabasePath.isEmpty()) {
```

Why only if it's empty? I'd be tempted to migrate these regardless.  And it's possible that some OS or file system combinations will wind up automatically creating a file in a new directory (e.g. MacOS `.DS_Store` cruft), not normal but we wouldn't want to silently skip copying the files.

---

this is primarily for testing, it could be left out.
basically at one point i was calling migrate and not the sub functions at the wrong time so this is probably safe to remove now as the tests do either/or not both now.